<a href="https://colab.research.google.com/github/theventurecity/data-toolkit/blob/master/Full_Pipeline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![TheVentureCity](https://theventure.city/wp-content/uploads/2017/06/Theventurecity-logoweb-1.png)

# Full Pipeline
1. Extract raw event log data from a CSV
1. Transform the raw data into analysis dataframes
    - Growth Accounting
    - Cohort Analysis
    - Engagement
1. Load the transformed data into Google Sheets 
1. Visualize insights in Google Data Studio

## Before you begin

- This notebook is shared in Google Colaboratory with read-only access. To run this notebook yourself, first click "**Open in Playground**" in the toolbar above. That will create a separate instance that you can run and/or save a copy of to your own Google Drive. 

- To run each cell, hit **Shift-Enter**, which will run the contents of the active cell and move to the next cell. This includes the markup cells (such as this one).

- When you run the first block of Python code, you will get a message that says, "**Warning: This notebook was not authored by Google.**" Please be aware that we are **NOT** accessing your data shared with Google or reading data and credentials from other sessions. This notebook reads data from GitHub and writes to a Google Sheet that only you have access to and can control. We recommend you click the box to "**Reset all runtimes before running**" for extra information security.

## Import relevant Python libraries

In [1]:
### Pandas to read the data from its source and manipulate it in memory
import pandas as pd


### The IPython.display library allows us to embed an iFrame within this 
### notebook
from IPython.display import IFrame


### To run this with functions from TheVentureCity's GitHub repository,
### clone the repository to the Google Colaboratory runtime environment
### and then import the code. This code allows us to run pre-existing functions 
### rather than having to define them inline within the notebook
### THIS IS ONLY APPLICABLE IF YOU WANT TO ACCESS THOSE FUNCTIONS ###
from importlib.machinery import SourceFileLoader
!git clone https://github.com/theventurecity/data-toolkit.git /tmp/theventurecity
!mv /tmp/theventurecity/python/tvc_transform.py tvc_transform.py
!mv /tmp/theventurecity/python/tvc_load_colab.py tvc_load_colab.py
!rm -r /tmp/theventurecity
tvct = SourceFileLoader('tvc_transform', 'tvc_transform.py').load_module()
tvcl = SourceFileLoader('tvc_load_colab', 'tvc_load_colab.py').load_module()

Cloning into '/tmp/theventurecity'...
remote: Enumerating objects: 157, done.
remote: Counting objects: 100% (157/157), done.
remote: Compressing objects: 100% (127/127), done.
remote: Total 157 (delta 84), reused 84 (delta 29), pack-reused 0
Receiving objects: 100% (157/157), 37.97 MiB | 17.65 MiB/s, done.
Resolving deltas: 100% (84/84), done.


## 1. Extract raw event log data from a CSV

This example uses a data file for a sample company from our GitHub repository called ServBiz. In this step we read the data file into memory as a Pandas dataframe we name "t."

In [2]:
filename = 'https://raw.githubusercontent.com/theventurecity/analytics/master/data/ServBiz_transactions.csv'
t = pd.read_csv(filename)
t.tail(10)

client_id        date  value_usd     segment
420781    27902A  2019-02-28       8.75  Enterprise
420782    34181A  2019-02-28      18.97         SMB
420783    30168A  2019-02-28      17.73         SMB
420784    30844A  2019-02-28      19.98         SMB
420785    35815A  2019-02-28      17.98         SMB
420786    16958A  2019-02-28      17.45         SMB
420787    13090A  2019-02-28      13.48         SMB
420788    19162A  2019-02-28      13.64  Enterprise
420789    28409A  2019-02-28      14.72         SMB
420790    12080A  2019-02-28      18.32         SMB

## 2. Transform the raw data into analysis dataframes
### 2.1 Create DAU Decorated and MAU Decorated "building block" dataframes
**Note**: For a more detailed discussion about creating the DAU and DAU Decorated dataframes, complete with inline code, visit [Create the DAU Decorated Data Set](https://colab.research.google.com/drive/12uehG2EcIqxcTazKs-pNQRTQSckllOmE)
#### 2.1.1 Create Daily Active Users (DAU) dataframe
The **DAU** dataframe aggregates all activity by user and day. 

In [3]:
# Run the create_dau_df function and show the first ten rows of the resulting dataframe
dau = tvct.create_dau_df(t, 
                         user_id = 'client_id', 
                         activity_date = 'date', 
                         inc_amt = 'value_usd',
                         segment_col = 'segment'
                        )
dau.head(10)

user_id activity_date     segment  inc_amt
0  10000A    2015-10-14         SMB    11.75
1  10001A    2015-10-14         SMB    13.75
2  10001A    2015-11-02         SMB     7.50
3  10001A    2015-11-22         SMB    18.00
4  10001A    2017-04-04         SMB     6.25
5  10001A    2017-12-08         SMB     8.75
6  10002A    2015-10-14  Enterprise    11.75
7  10002A    2015-10-26  Enterprise    12.25
8  10002A    2015-11-16  Enterprise    12.25
9  10002A    2015-11-23  Enterprise    12.25

#### 2.1.2 Calculate First Date and DAU Decorated dataframes
The create_dau_decorated_df function calls the create_first_dt_df if no first_dt dataframe is specified

In [4]:
# Run the create_dau_decorated_df function and show the first ten rows of the resulting dataframe
dau_decorated = tvct.create_dau_decorated_df(dau)
dau_decorated.head(10)

Creating DAU Decorated dataframe
Creating first_dt dataframe


user_id activity_date     segment  inc_amt    first_dt  \
0  10000A    2015-10-14         SMB    11.75  2015-10-14   
1  10001A    2015-10-14         SMB    13.75  2015-10-14   
2  10001A    2015-11-02         SMB     7.50  2015-10-14   
3  10001A    2015-11-22         SMB    18.00  2015-10-14   
4  10001A    2017-04-04         SMB     6.25  2015-10-14   
5  10001A    2017-12-08         SMB     8.75  2015-10-14   
6  10002A    2015-10-14  Enterprise    11.75  2015-10-14   
7  10002A    2015-10-26  Enterprise    12.25  2015-10-14   
8  10002A    2015-11-16  Enterprise    12.25  2015-10-14   
9  10002A    2015-11-23  Enterprise    12.25  2015-10-14   

             first_week first_month  
0 2015-10-12/2015-10-18     2015-10  
1 2015-10-12/2015-10-18     2015-10  
2 2015-10-12/2015-10-18     2015-10  
3 2015-10-12/2015-10-18     2015-10  
4 2015-10-12/2015-10-18     2015-10  
5 2015-10-12/2015-10-18     2015-10  
6 2015-10-12/2015-10-18     2015-10  
7 2015-10-12/2015-10-18     2015-10  
8 2015-10-12/2015-10-18     2015-10  
9 2015-10-12/2015-10-18     2015-10

Combining the basic DAU data with the first date, week, and month for each user, **the DAU Decorated dataframe is our basic building block for many different analyses**. It allows us to use user-level data to inspect engagement, retention, and growth accounting.
#### 2.1.3 Calculate MAU Decorated dataframe

In [5]:
# Run create_xau_decorated_df for MAU's (using 'month'), unsegmented
mau_decorated = tvct.create_xau_decorated_df(dau_decorated, 'month', use_segment=False)
mau_decorated.tail(10)

Creating Monthly Active Users Decorated dataframe


Month_Year user_id  inc_amt first_month Next_Month_Year
93658    2019-02   9784A   533.66     2015-09         2019-03
93659    2019-02   9794A    24.72     2015-09         2019-03
93660    2019-02   9808A    85.12     2015-09         2019-03
93661    2019-02   9868A    65.75     2015-10         2019-03
93662    2019-02   9876A    19.48     2015-10         2019-03
93663    2019-02   9902A    39.76     2015-10         2019-03
93664    2019-02   9952A    53.50     2015-10         2019-03
93665    2019-02   9986A   116.64     2015-10         2019-03
93666    2019-02   9989A    21.28     2015-10         2019-03
93667    2019-02   9995A    85.12     2015-10         2019-03

### 2.2 Transform into a Growth Accounting analysis dataframe
**Note**: The function called in this section is different than what appears in the [Mini-Pipeline: Growth Accounting](https://colab.research.google.com/drive/1moHa4Mcycwsz7Fq6T_5Zou1Zunt0afiI), which was condensed for brevity. The resultant dataset is therefore adjusted to write only the relevant columns to Google Sheets.

In [6]:
TIME_PER = 'month'
USE_SEGMENT = False
KEEP_LAST_PER = True
DATE_LIMIT = None
INCL_ZERO_INC = False

user_ga, rev_ga = tvct.create_growth_accounting_dfs(mau_decorated, 
                                                    time_period=TIME_PER, 
                                                    use_segment=USE_SEGMENT, 
                                                    keep_last_period=KEEP_LAST_PER, 
                                                    date_limit=DATE_LIMIT, 
                                                    include_zero_inc=INCL_ZERO_INC)
user_ga_with_ratios = tvct.calc_user_ga_ratios(user_ga, 
                                               time_period=TIME_PER, 
                                               use_segment=USE_SEGMENT, 
                                               growth_rate_periods=12)
user_ga_with_ratios.tail(10)

Creating Growth Accounting dataframes


Month_Year  Monthly Active Users  Retained Users  New Users  \
36    2018-05                  3253            2507        441   
37    2018-06                  3183            2626        340   
38    2018-07                  3377            2691        386   
39    2018-08                  3570            2835        453   
40    2018-09                  3561            2947        383   
41    2018-10                  3972            3034        581   
42    2018-11                  4102            3307        518   
43    2018-12                  4522            3482        740   
44    2019-01                  4689            3811        583   
45    2019-02                  4657            3865        537   

    Resurrected Users  Churned Users segment  Users BOP  MoM User Retention  \
36                305           -571     All     3078.0            0.814490   
37                217           -627     All     3253.0            0.807255   
38                300           -492     All     3183.0            0.845429   
39                282           -542     All     3377.0            0.839503   
40                231           -623     All     3570.0            0.825490   
41                357           -527     All     3561.0            0.852008   
42                277           -665     All     3972.0            0.832578   
43                300           -620     All     4102.0            0.848854   
44                295           -711     All     4522.0            0.842769   
45                255           -824     All     4689.0            0.824270   

    User Quick Ratio  User CMGR12  Growth Threshold  User CMGR12 Target  
36          1.306480     0.032119               1.0                 0.1  
37          0.888357     0.025316               1.0                 0.1  
38          1.394309     0.033163               1.0                 0.1  
39          1.356089     0.036383               1.0                 0.1  
40          0.985554     0.036500               1.0                 0.1  
41          1.779886     0.038371               1.0                 0.1  
42          1.195489     0.036690               1.0                 0.1  
43          1.677419     0.042951               1.0                 0.1  
44          1.234880     0.040467               1.0                 0.1  
45          0.961165     0.038370               1.0                 0.1

### 2.3 Transform into a Cohort Analysis dataframe
Create the MAU Cohorts dataframe, using MAU Decorated as an input

In [7]:
mau_cohorts = tvct.create_xau_cohort_df(mau_decorated, 'month')
mau_cohorts.tail(10)

first_month Month_Year  Months Since First   inc_amt  cust_ct  \
1071     2018-11    2018-11                   0  17178.12      518   
1072     2018-11    2018-12                   1  20136.67      356   
1073     2018-11    2019-01                   2  18325.46      300   
1074     2018-11    2019-02                   3  14171.36      263   
1075     2018-12    2018-12                   0  26134.01      740   
1076     2018-12    2019-01                   1  31004.43      497   
1077     2018-12    2019-02                   2  22879.87      379   
1078     2019-01    2019-01                   0  20557.47      583   
1079     2019-01    2019-02                   1  21454.77      388   
1080     2019-02    2019-02                   0  17109.64      537   

      cohort_cust_ct  cum_inc_amt  cum_inc_per_cohort_cust  cust_ret_pct  
1071             518     17178.12                33.162394      1.000000  
1072             518     37314.79                72.036274      0.687259  
1073             518     55640.25               107.413610      0.579151  
1074             518     69811.61               134.771448      0.507722  
1075             740     26134.01                35.316230      1.000000  
1076             740     57138.44                77.214108      0.671622  
1077             740     80018.31               108.132851      0.512162  
1078             583     20557.47                35.261527      1.000000  
1079             583     42012.24                72.062161      0.665523  
1080             537     17109.64                31.861527      1.000000

### 2.4. Transform into Engagement dataframes
#### 2.4.1 Calculate User Periodic Usage
Now that we have the "DAU Decorated" data frame, we can use it to calculate engagement metrics since they are super-important for an early-stage startup. The **calc_user_periodic_usage** function below calculates the number of active days for each user in a 28-day window of time. It outputs a dataframes that holds the active day count for every user, sorted by their inc_amt.

In [0]:
### Run calc_user_periodic_usage and display the first 10 rows
#user_daily_usage = tvct.calc_user_periodic_usage(dau_decorated, 
#                                            'day', 
#                                            dau_decorated['activity_date'].max(), 
#                                            window_days=28, 
#                                            breakouts=[2, 4], 
#                                            use_segment=False
#                                           )
#user_daily_usage.head(10)

The calc_user_periodic_usage function above serves as a key helper function for the functions below that summarize its data into the data that will be used for the visualizations.

#### 2.4.2 Calculate the DAU Histogram
The DAU histogram puts each user into a bin according to the number of active days they have in the 28-day window. It outputs the user count in each bin and also counts the average number of days active across all users during the window.

In [9]:
dau_hist_L28 = tvct.calc_xau_hist(dau_decorated, 
                                  'day', 
                                  dau_decorated['activity_date'].max(), 
                                  window_days=28, 
                                  use_segment=False)
dau_hist_L28

active_days_bin  user_count  avg_days_active
0                1         921         4.072579
1                2        1006         4.072579
2                3         487         4.072579
3                4        1148         4.072579
4                5         104         4.072579
5                6         165         4.072579
6                7         135         4.072579
7                8         305         4.072579
8                9          52         4.072579
9               10         116         4.072579
10              11          30         4.072579
11              12          30         4.072579
12              13           9         4.072579
13              14          19         4.072579
14              15           7         4.072579
15              16          14         4.072579
16              17          15         4.072579
17              18          60         4.072579
18              19           6         4.072579
19              20           3         4.072579
20              21           4         4.072579
21              22          21         4.072579
22              23           0         4.072579
23              24           0         4.072579
24              25           0         4.072579
25              26           0         4.072579
26              27           0         4.072579
27              28           0         4.072579

#### 2.4.3 Calculate Rolling DAU/MAU Ratios over Time
The DAU Histogram shown above helps us understand engagement during a fixed 28-day period. What if we could calculate its key attributes to look at engagement trends over time? We can do so by performing a similar analysis and looping through it on a daily basis. In doing so, we can build out a dataframe that contains engagement metrics for every 28-day period over the life of the company.

In [10]:
# Run rolling_dau_mau and show the last ten rows (the most recent ones)
rolling_dau_mau = tvct.create_xau_window_df(dau_decorated, breakouts = [2, 4, 8, 12])
rolling_dau_mau.tail(10)

1372 total days to process...
Processing day 0 of 1372...
Processing day 100 of 1372...
Processing day 200 of 1372...
Processing day 300 of 1372...
Processing day 400 of 1372...
Processing day 500 of 1372...
Processing day 600 of 1372...
Processing day 700 of 1372...
Processing day 800 of 1372...
Processing day 900 of 1372...
Processing day 1000 of 1372...
Processing day 1100 of 1372...
Processing day 1200 of 1372...
Processing day 1300 of 1372...
Finished processing all 1372 days!


index  active_days  1d+ users  dau_window_ratio  window_frequency  \
0      0        18775       4690          0.142971          4.003198   
0      0        18789       4694          0.142956          4.002769   
0      0        18821       4695          0.143169          4.008733   
0      0        18844       4709          0.142918          4.001699   
0      0        18853       4712          0.142895          4.001061   
0      0        18871       4697          0.143488          4.017671   
0      0        18865       4691          0.143626          4.021531   
0      0        18868       4684          0.143864          4.028181   
0      0        18901       4675          0.144393          4.042995   
0      0        18966       4657          0.145449          4.072579   

   2d+ users  2d+ users / total 28d users  4d+ users  \
0       3719                     0.792964       2185   
0       3711                     0.790584       2187   
0       3715                     0.791267       2200   
0       3720                     0.789977       2214   
0       3722                     0.789898       2216   
0       3722                     0.792421       2221   
0       3724                     0.793861       2216   
0       3719                     0.793980       2219   
0       3728                     0.797433       2229   
0       3736                     0.802233       2243   

   4d+ users / total 28d users  8d+ users  8d+ users / total 28d users  \
0                     0.465885        682                     0.145416   
0                     0.465914        681                     0.145079   
0                     0.468584        677                     0.144196   
0                     0.470164        678                     0.143980   
0                     0.470289        678                     0.143888   
0                     0.472855        680                     0.144773   
0                     0.472394        674                     0.143679   
0                     0.473740        680                     0.145175   
0                     0.476791        681                     0.145668   
0                     0.481641        691                     0.148379   

   12d+ users  12d+ users / total 28d users window_end_dt  
0         186                      0.039659    2019-02-19  
0         186                      0.039625    2019-02-20  
0         186                      0.039617    2019-02-21  
0         186                      0.039499    2019-02-22  
0         186                      0.039474    2019-02-23  
0         188                      0.040026    2019-02-24  
0         187                      0.039864    2019-02-25  
0         186                      0.039710    2019-02-26  
0         188                      0.040214    2019-02-27  
0         188                      0.040369    2019-02-28

## 3. Load the transformed data into Google Sheets 
**Note**: For a more detailed discussion about loading data into Google Sheets, complete with inline code, visit [Section 3 of Mini-Pipeline: Cohort Analysis](https://colab.research.google.com/drive/1oYy-wJl6VZFgOsv8uw7iGChQxUjrR5rf#scrollTo=hzG-fUwkyM9K)
### 3.1 Establish connection to Google Sheets for writing output files
The first time you run the cell below, or after some time of inactivity, you will be asked to click on a link. That link will take you to a new tab that will authorize this script to write to Google Sheets spreadsheets in your Google Account. To enable this feature, copy the code you get into the box below and hit Enter.

In [0]:
gc = tvcl.google_authenticate()

In [0]:
### Be sure to set this value to refer to your Google Sheets workbook
GOOGLE_SHEET_KEY = '1-XnO_eWkRwX-E1fiA2Jkbe3kJvoyoPFsdeW7vnF6zS0' 

In [13]:
print('https://docs.google.com/spreadsheets/d/' + GOOGLE_SHEET_KEY)

https://docs.google.com/spreadsheets/d/1-XnO_eWkRwX-E1fiA2Jkbe3kJvoyoPFsdeW7vnF6zS0


### 3.2 Write the output files to Google Sheets

In [0]:
### Execute this function to write the data in the dataframe to the google sheet
### and tab name specified using the gc Google credentials
tvcl.write_to_google_sheet(user_ga_with_ratios, 
                           GOOGLE_SHEET_KEY, 
                           'MAU Growth Accounting', 
                           gc)

In [0]:
### Execute this function to write the data in the dataframe to the google sheet
### and tab name specified using the gc Google credentials
tvcl.write_to_google_sheet(mau_cohorts, 
                           GOOGLE_SHEET_KEY, 
                           'MAU Retention by Cohort', 
                           gc)

In [0]:
### Execute this function to write the data in the dataframe to the google sheet
### and tab name specified using the gc Google credentials
tvcl.write_to_google_sheet(dau_hist_L28, 
                           GOOGLE_SHEET_KEY, 
                           'DAU Histogram L28', 
                           gc)

In [0]:
### Execute this function to write the data in the dataframe to the google sheet
### and tab name specified using the gc Google credentials
tvcl.write_to_google_sheet(rolling_dau_mau, 
                           GOOGLE_SHEET_KEY, 
                           'Rolling DAU/MAU', 
                           gc)

## 4. Visualize insights in Google Data Studio
A Google DataStudio dashboard preconfigured to read from the Google Sheet created above to visualize the data is [available at this link](https://datastudio.google.com/open/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV) or by clicking on the Google DataStudio logo at the bottom of the chart embedded below. It is available in read-only mode for you to copy, link to your own Google Sheet tabs, and see your own data visualized.
### 4.1 Show MAU
The first chart shows MAU with the three main active categories broken out:

- Retained Users are in blue
- New Users are in light green
- Resurrected Users are in darker green

Together these three categories add up to the total MAU for each month. The line snaking through the chart is the compound monthly growth rate calculated over a 12-month period (CMGR12). We see that it starts off high when the growth starts from a low base. But over time, the CMGR12 stabilizes around 4%. The User GMGR12 Target line of 10% serves as an example of a healthy growth rate for a successful startup, though this can vary depending on the situation.

In [23]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/YVKk', 
       width=600, 
       height=450)

### 4.2 Show MAU Growth Accounting
The chart below is similar to the last one in that it still shows the same New and Resurrected user counts in light green and darker green, respectively. This chart removes the retained user count in favor of churned users in orange (expressed as a negative number). It also removes the CMGR12 line in favor of two new lines: the User Quick Ratio (in black) and MoM User Retention rate (in blue). 

- User Quick Ratio = -1 * (New Users + Resurrected Users) / Churned Users
  - In other words, it is the ratio of the green bars at the top and the orange bar on the bottom
  - The horizontal line where the Quick Ratio = 1 serves as the demarcation line between growth and shrinkage
  - A Quick Ratio > 1 means that MAUs grew from last month to this month
  - A Quick Ratio < 1 means that MAUs shrank from last month to this month
  - For more about the Quick Ratio, visit our blog post [Quick Ratio as a Shortcut to Understand Product Growth](https://medium.com/theventurecity/quick-ratio-as-a-shortcut-to-understand-product-growth-ae60212bd371)
- MoM Retention Rate = Retained Users / Users BOP
  - Users BOP ("beginning of period") are the same as last month's users as calculated in the calc_user_ga_ratios function

In [24]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/7eKk', 
       width=600, 
       height=450)

### 4.3 Monthly Cohort User Retention
One way to think about cohort user retention is based on the fact that, typically, not all users are retained from Month 0 to Month 1, and then to Month 2, 3, 4, etc. The chart below depicts this in the downward curve approaching. If you mouseover the bars, we can see that the January 2018 cohort at the farthest right is at 23.7% user retention in Month 13. Compare that to the 60.94% figure for that same cohort in Month 1.

Sometimes this data is depicted as downward curving lines, one for each cohort. We prefer this visual because (a) the lines often end up looking jumbled and spaghetti-like; and (b) it allows us to see the trend in each Months Since First's retention. For example, the most recent Month 1 retention of 66.55% for the January 2019 cohort is down from the Month 1 peak of 72.8% percent for the July 2018 cohort. In this way, we can see the general downward slope that the spaghetti lines would give us, but also see the trends within each Months Since First, for a superior visual.

In [25]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/EOyj', 
       width=600, 
       height=450)

### 4.4 Cumulative Customer Revenue LTV by Monthly Cohort
The visual below shows the cumulative inc_amt *per original cohort customer* that each cohort has generated in the time it has been active. These lines give a look at how well a group of ServBiz's users keeps generating revenue after they are initially acquired. This gives us an empirical way to look at customer long-term value (LTV) over a range of possible scenarios. For example, if you mouseover 16 Months Since First in the chart below, the four cohorts that have been around that long range from `$`440.47 to `$`588.87 per cohort customer. We can use that as a 16-month LTV range for projection or for calculating LTV-to-CAC ratio.

Another feature of the curves below is that, while they are slighly bending downward, the rate of change in the slope is very gradual. This implies that ServBiz's customers are active for a long period of time, meaning that the time horizon for calculating LTV may be extended and that these customers are quite value in the long-term.

In [26]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/pVBk', 
       width=600, 
       height=450)

### 4.5 User DAU Histogram
The chart below shows the number of users in each active days bin to illustrate the distribution of usage. The more users come back to the service on multiple days, the more engaged they are. We see that the highest bar is at 4 active days, or once per week. We also see another, smaller spike at 8 active days, or twice per week. The 4.1 figure for "Average Days Active in Last 28" is another way of expressing the DAU/MAU ratio. The DAU/MAU ratio in this case is 4.1/28, or 0.146, or 14.6%. That means users are active on 14.6% percent of possible days. We think it is easier to understand what this number means by thinking of it as average active days out of 28 rather than as a ratio/percentage.

In [19]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/IEWi', 
       width=600, 
       height=450)

### 4.6 Rolling Active Days per 28
The chart below takes the 4.1 Average Active Days in Last 28 from the last chart--which was calculated from the most recent date in the event log, and plots the trend of that number over time. By looking at this number's trend, we can judge whether engagement is getting better, getting worse, or staying the same. In the case of ServBiz, engagement has leveled out around 4.0 over the last year-and-a-half or so after showing steady improvement in the first few years of the business. The exception is the regular seasonal dips when usage frequency wanes for a few weeks.

In [20]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/hXXk', 
       width=600, 
       height=450)

### 4.7 Rolling 28-Day Multi-Day Users
A richer display of the distibution of active usage in effect combines the histogram from 4.1 above with the rolling trends from 4.2. In effect, we are turning the histogram on its side and dragging it through time to see the percentage of total active users above various thresholds. In the example below we have plotted four lines. They show the percentage of active users with...

- 2+ active days out of the previous 28, or once at least every other week (about 80% in the most recent 28-day window; see table below the chart)
- 4+ active days, or at least once a week (about 48%)
- 8+ active days, or at least twice a week (about 15%)
- 12+ active days, or at least 3x per week (about 4%)

Going with more than 12 days would make for a lousy visual with a horizontal line bouncing off zero.

In this visual we see that the percentage of users with 2+ days out of 28 continues to increase a bit over time. It is the more frequent categories (4+, 8+) that are flat and dragging down the overall average seen in 4.2.

In [21]:
IFrame('https://datastudio.google.com/embed/reporting/1xjS__Q6ZUXuUUARkgRvY4spYUw1ePksV/page/qdXk', 
       width=600, 
       height=450)

In [22]:
rolling_dau_mau.tail(1)

index  active_days  1d+ users  dau_window_ratio  window_frequency  \
0      0        18966       4657          0.145449          4.072579   

   2d+ users  2d+ users / total 28d users  4d+ users  \
0       3736                     0.802233       2243   

   4d+ users / total 28d users  8d+ users  8d+ users / total 28d users  \
0                     0.481641        691                     0.148379   

   12d+ users  12d+ users / total 28d users window_end_dt  
0         188                      0.040369    2019-02-28